Add TypeScript typings for functional components and inclusive components

jmrog opened the pull request:
This PR contains a few more proposed improvements to Preact's TypeScript declarations, related tangentially to some things mentioned in PR #573. In #573, I mentioned that the compiler complains at times when you do things with higher-order components, like so:

```typescript
function getName(component: typeof preact.Component): string {
    return component.displayName || component.name || 'Component';
}
```

The issue in #573 was with the `displayName` property. But _this_ PR addresses the fact that there also appears to be no out-of-the-box way to do similar things with _functional_ Preact components as opposed to "full-fledged" components (i.e., components with the full lifecycle implementation). That is, there seems to be no _out-of-the-box_ way (i.e., without creating my own custom types) to do something like this:

```typescript
// I avoided the name "GenericComponent" because the term "generic" is overloaded in the context of Typescript, so "InclusiveComponent" it is.
function getName(component: preact.InclusiveComponent<any, any>): string {
    return component.displayName || component.name || 'Component';
}
```

or even like this:

```typescript
function getName(component: typeof preact.Component | preact.FunctionalComponent<any>): string {
    return component.displayName || component.name || 'Component';
}
```

Ideally, I would be able to specify that a higher-order function accepts Preact components generally, whether those components are mere functional components or are "full-fledged" components, as in the first of the two preceding examples (without having to create my own types, since functional components are standard to the ecosystem and are widely used interchangeably with full-fledged components). This PR lets you do that, with type-checking/without the TypeScript compiler complaining.

robertknight commented on src/preact.d.ts:
```diff
@@ -37,6 +37,16 @@ declare namespace preact {
 		new (props?:PropsType):Component<PropsType, StateType>;
 	}
 
+	interface FunctionalComponent<PropsType> {
```
In the React typings, the equivalent interface is called `StatelessComponent`. Using the same name would probably be helpful to people migrating from one to the other. Thoughts?

developit replied:
This makes sense to me

robertknight commented on src/preact.d.ts:
```diff
@@ -37,6 +37,16 @@ declare namespace preact {
+
+	// An "inclusive" component is a component that is either a functional
+	// component or a more full-fledged component.
+	type InclusiveComponent<PropsType, StateType> = FunctionalComponent<PropsType> | typeof Component;
```
Hmm ... this seems like a useful alias to provide but the name sounds a bit weird. I'm struggling to think of something much better. `AnyComponent`?

jmrog commented:
Agreed on all counts. Updated PR.

robertknight commented:
@jmrog - Thanks for the update. Separately I merged https://github.com/developit/preact/pull/605 and forgot that this PR implemented very similar changes. Could you have a look at the current code in master and let me know whether there is anything from this PR that still needs to be added?

jmrog commented:
@robertknight Thanks for letting me know. Sure, I'll take a look and update as necessary, once time allows (probably very soon).

jmrog commented:
@robertknight: I updated this PR again to reflect the new material on master, though I do have one lingering question now. Earlier in this thread, you suggested renaming `FunctionalComponent` to `StatelessComponent` in order to follow the React naming conventions more closely. I did that; however, the other PR (which was merged) uses the name `FunctionalComponent`. So, for the moment, I've switched back to using `FunctionalComponent` for this PR. If `StatelessComponent` is still actually preferred, I can flip it back to that again. Is there a preference for that anymore?

The other remaining changes in this PR are small: 1) `displayName` and `defaultProps` are added as recognized properties for the `FunctionalComponent` type; 2) the `AnyComponent` type (a component that is either a functional component or a "class-y" component) is provided for convenience, since it will often be helpful for higher-order components; 3) code formatting is made more consistent in a few places.

Thanks!

robertknight commented:
> If StatelessComponent is still actually preferred, I can flip it back to that again. Is there a preference for that anymore?

No. I noticed that the [React docs](https://facebook.github.io/react/docs/components-and-props.html) use the term "functional component", so it is perhaps better to use that term even if it differs from the React typings.

Thanks for the contribution 👍
